Source/Core/Core/HW/DSPHLE/UCodes/ASnd.cpp

```diff
@@ -0,0 +1,544 @@
+
+        const auto [new_r, new_l] = (this->*sample_function)();
+        // out_samp: "multiply sample x volume" - left is put in $ax0.h, right is put in $ax1.h
+
+        // All functions jumped to from sample_selector jump or fall through here (zero_samples also
+        // jumps here, but is handled separately above for structural reasons)
+        s32 temp_l = static_cast<s32>(new_l) * m_current_voice.volume_l;
+        m_current_voice.left = static_cast<s16>(temp_l >> 8);
+        s32 temp_r = static_cast<s32>(new_r) * m_current_voice.volume_r;
+        m_current_voice.right = static_cast<s16>(temp_r >> 8);
+      }
+      // loop_end, remember that the labeled instruction is the last instruction in the loop
+      // so jumping to the NOP here causes the loop to repeat (and decrement its counter)
+    }
+    // end_process (not actually jumped to)
+    if (m_current_voice.start_addr == 0)
+    {
+      return_mail = DSP_SYNC;
+      ChangeBuffer();
+    }
+    // save_datas_end (we mutate m_current_voice.start_addr, left, and right in place,
+    // so nothing needs to happen here here)
+  }
+end_main:  // end_main
+  DMAOutVoiceData();
+
+  m_mail_handler.PushMail(return_mail, true);
+}
+
+void ASndUCode::ChangeBuffer()
+{
+  // change_buffer
+  m_current_voice.volume_l = m_current_voice.volume2_l;
+  m_current_voice.volume_r = m_current_voice.volume2_r;
+  m_current_voice.end_addr = m_current_voice.end_addr2;
+  m_current_voice.start_addr = m_current_voice.start_addr2;
+  m_current_voice.backup_addr = m_current_voice.start_addr2;
+
+  const u32 loop_mask =
+      (m_crc == HASH_2011 || m_crc == HASH_2020) ? NEW_FLAGS_VOICE_LOOP : OLD_FLAGS_VOICE_LOOP;
+
+  if ((m_current_voice.flags & loop_mask) == 0)
+  {
+    m_current_voice.start_addr2 = 0;
+    m_current_voice.end_addr2 = 0;
+  }
+}
+
+void ASndUCode::DMAInSampleData()
+{
+  // load_smp_addr_align
+  // This is its own function, and it contains its own copy of the DMA logic.
+  // The only difference is that this one forces the address to be aligned, while when
+  // jump_load_smp_dma is used, the address is expected to already be aligned.
+  const u32 addr = m_current_voice.start_addr & ~0x1f;
+  for (u16 i = 0; i < 0x10; i++)
+  {
+    m_input_sample_buffer[i] = HLEMemory_Read_U16(addr + i * sizeof(u16));
+  }
+}
+
+void ASndUCode::DMAInSampleDataAssumeAligned()
+{
+  // jump_load_smp_dma
+  // This is technically not a function, but instead is directly jumped to and then jumps to $ar3
+  // (which is set to an address from sample_selector). We can just treat it as a function though.
+  const u32 addr = m_current_voice.start_addr;
+  for (u16 i = 0; i < 0x10; i++)
+  {
+    m_input_sample_buffer[i] = HLEMemory_Read_U16(addr + i * sizeof(u16));
+  }
+}
+
+std::pair<s16, s16> ASndUCode::ReadSampleMono8Bits() const
+{
+  // mono_8bits
+  s16 result = m_input_sample_buffer[(m_current_voice.start_addr >> 1) & 0xf];
```

Comment:
Maybe no particular reason but is there any significance of the `0xf` that I see in all this code?

Reply:
m_input_sample_buffer is 16 entries long, so we mask with 15 to theoretically keep it in-bounds. (`ReadSampleStereo16Bits` adds 1 after masking, which means it could go out of bounds if `start_addr` is not divisible by 4, but this is also the case with the real ucode... not 100% sure what to do with that.)

As for *why* it's 16, I think the DSP DMA requires 32-byte alignment... or at least the card asnd ucode was written with that assumption.  (16 two-byte values is 32 bytes.)

Reply:
I've created constants for these various masks, and also increased the size of `m_input_sample_buffer` to 32 words (which is the case in the actual uCode, even though it does mask at 15) which resolves the out-of-bounds index edge case.

Reply:
Awesome, it looks much nicer.  Thanks!